src/components/calcite-ui-icon/utils.ts

```diff
@@ -0,0 +1,69 @@
+import { getAssetPath } from "@stencil/core";
+import { Scale } from "../../interfaces/UIIcon";
+
+export interface FetchIconProps {
+  icon: string;
+  scale: Scale;
+  filled: boolean;
+}
+
+/**
+ * Icon data cache.
+ * Exported for testing purposes.
+ * @private
+ */
+export const iconCache: Record<string, string> = {};
+
+/**
+ * Icon request cache.
+ * Exported for testing purposes.
+ * @private
+ */
+export const requestCache: Record<string, Promise<any>> = {};
+
+export const scaleToPx: Record<Scale, number> = {
+  s: 16,
+  m: 24,
+  l: 32
+};
+
+export async function fetchIcon({
+  icon,
+  scale,
+  filled
+}: FetchIconProps): Promise<string> {
+  const size = scaleToPx[scale];
+  const id = `${normalizeIconName(icon)}${size}${filled ? "F" : ""}`;
+
+  if (iconCache[id]) {
+    return iconCache[id];
+  }
+
+  const request =
+    requestCache[id] ||
+    (requestCache[id] = import(getAssetPath(`./assets/${id}.js`)));
+
+  const module = await request;
+  const pathData = module[id];
+
+  iconCache[id] = pathData;
+  return pathData;
+}
+
+function normalizeIconName(name: string): string {
+  const parts = name.split("-");
+
+  if (parts.length === 0) {
```

Comment:
I might not understand correctly, but this code might be doing nothing. Any string you split will be a length of at least 1 regardless of what you split on, yeah?

ie:

```
"".split("-") => [""] 
```